### ISSUE TITLE
Scripted Dashboards

### ISSUE THREAD

**Issue by sabino** [metabase/metabase#9434]
### Feature requests and proposals

I think it should be possible to create scripted dashboards. Something like this: http://docs.grafana.org/reference/scripting/

I know that it is possible to use the API but this is different. The dashboard stuff would not be store in the database, but maybe the filesystem. (I was thinking maybe something similar to the [LookML stuff from Looker](https://docs.looker.com/dashboards/creating-lookml-dashboards)?). It would get stored in the database after importing.

And maybe the dashboard could be created based on the parameters coming from the files and they can't be edited unless you "reimport", or change the script that generated them.

That way it would be possible to recreate the dashboards anytime, and use an already known version control system, like git.

**Comment by sabino** [comment]
I just realized that this could be an extension for the X-rays / Automagic Dashboards.
The way it works today is that we can create a dashboard out of it, but there is no way to update an existing dashboard based on a `yaml` change right?

It could also be possible to point a git repo to link, so all `yaml` file would be placed there and there would be no need to deploy metabase on every new automagic table definition.

**Comment by sbelak** [comment]
No, it's not, once you save an xray it "forgets" that it was an xray and acts just like any other dashboard. 

One thing you could try is to use the `POST /api/dashboard/save` endpoint (which xrays use when saving) which takes as input a complete dashboard description in JSON. This will create a new dashboard though which might be awkward as the urls will change and you'd have to delete the old instance. 

What exactly is your use-case? Maybe there are also some more partial solutions.

**Comment by sabino** [comment]
A use case would be to keep a reproducible version of the dashboard somewhere safe, and another is to make possible to review the dashboard before "going to production".

One could have a metabase running locally to develop a dashboard and after that export the settings of this dashboard an import into the shared production server. Also we could keep the server always clean as it could be easily redeployed from scratch and all the dashboards/questions would be there.

**Comment by dragonsahead** [comment]
@sabino I would guess that you want something similar as https://www.metabase.com/docs/latest/configuring-metabase/config-file but for entities (dashboards/questions)

**Comment by sabino** [comment]
Yes, and it ended up being implemented after all.
It's called Serialization: https://www.metabase.com/docs/latest/installation-and-operation/serialization

The only drawback is that it's a paid feature.
